lib/virtio/virtqueue.c

```diff
@@ -129,7 +129,7 @@ int virtqueue_add_buffer(struct virtqueue *vq, struct virtqueue_buf *buf_list,
 
 	needed = readable + writable;
 
-	VQ_PARAM_CHK(vq == NULL, status, ERROR_VQUEUE_INVLD_PARAM);
+	VQ_PARAM_CHK(vq < needed, status, ERROR_VQUEUE_INVLD_PARAM);
```

Comment:
The comparison is not valid, you compare an integer with a structure pointer.
What about following check?
`VQ_PARAM_CHK((needed < 1 || needed > vq->vq_free_cnt), status, ERROR_VQUEUE_INVLD_PARAM);`

Reply:
I had actually fixed this as follows:

VQ_PARAM_CHK(vq->vq_free_cnt < needed, status, ERROR_VRING_FULL);

Not sure how I truncated vq_free_cnt on my commit.